crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+};
 
+use log::debug;
+use pipewire as pw;
+use pw::spa::param::ParamType;
+use pw::spa::pod::Pod;
+use std::ffi::CString;
+use std::mem::size_of;
+
+use pw::spa::sys::{
+    spa_audio_info_raw, SPA_AUDIO_CHANNEL_FC, SPA_AUDIO_CHANNEL_FL, SPA_AUDIO_CHANNEL_FR,
+    SPA_AUDIO_CHANNEL_LFE, SPA_AUDIO_CHANNEL_MONO, SPA_AUDIO_CHANNEL_RC, SPA_AUDIO_CHANNEL_RL,
+    SPA_AUDIO_CHANNEL_RR, SPA_AUDIO_CHANNEL_UNKNOWN, SPA_AUDIO_FORMAT_ALAW, SPA_AUDIO_FORMAT_F32,
+    SPA_AUDIO_FORMAT_F64, SPA_AUDIO_FORMAT_S16, SPA_AUDIO_FORMAT_S18_LE, SPA_AUDIO_FORMAT_S20,
+    SPA_AUDIO_FORMAT_S20_LE, SPA_AUDIO_FORMAT_S24, SPA_AUDIO_FORMAT_S24_LE, SPA_AUDIO_FORMAT_S32,
+    SPA_AUDIO_FORMAT_S8, SPA_AUDIO_FORMAT_U16, SPA_AUDIO_FORMAT_U18_LE, SPA_AUDIO_FORMAT_U20,
+    SPA_AUDIO_FORMAT_U20_LE, SPA_AUDIO_FORMAT_U24, SPA_AUDIO_FORMAT_U24_LE, SPA_AUDIO_FORMAT_U32,
+    SPA_AUDIO_FORMAT_U8, SPA_AUDIO_FORMAT_ULAW, SPA_AUDIO_FORMAT_UNKNOWN,
+};
+
+use pw::sys::{
+    pw_buffer, pw_loop, pw_thread_loop, pw_thread_loop_get_loop, pw_thread_loop_lock,
+    pw_thread_loop_new, pw_thread_loop_signal, pw_thread_loop_start, pw_thread_loop_unlock,
+    pw_thread_loop_wait, PW_ID_CORE,
+};
+use pw::{properties, spa, Context, Core, LoopRef};
+
+#[macro_export]
+macro_rules! properties_set {
+    ($properties:expr, $key:expr, $($args:tt)*) => {{
+        let key_cstr = CString::new($key).unwrap();
+        let format_str = format!($($args)*);
+        let format_cstr = CString::new(format_str).unwrap();
+        unsafe {
+            pw::sys::pw_properties_set(
+                $properties.as_ptr(),
+                key_cstr.as_ptr(),
+                format_cstr.as_ptr(),
+            )
+        }
+    }};
+}
+struct PwThreadLoop(NonNull<pw_thread_loop>);
+
+impl PwThreadLoop {
+    pub fn new(name: Option<&str>) -> Option<Self> {
+        let inner = unsafe {
+            pw_thread_loop_new(
+                name.map_or(ptr::null(), |p| p.as_ptr() as *const _),
+                std::ptr::null_mut(),
+            )
+        };
+        if inner.is_null() {
+            None
+        } else {
+            Some(Self(
+                NonNull::new(inner).expect("pw_thread_loop can't be null"),
+            ))
+        }
+    }
+
+    pub fn get_loop(&self) -> PwThreadLoopTheLoop {
+        let inner = unsafe { pw_thread_loop_get_loop(self.0.as_ptr()) };
+        PwThreadLoopTheLoop(NonNull::new(inner).unwrap())
+    }
+
+    pub fn unlock(&self) {
+        unsafe { pw_thread_loop_unlock(self.0.as_ptr()) }
+    }
+
+    pub fn lock(&self) {
+        unsafe { pw_thread_loop_lock(self.0.as_ptr()) }
+    }
+
+    pub fn start(&self) {
+        unsafe {
+            pw_thread_loop_start(self.0.as_ptr());
+        }
+    }
+
+    pub fn signal(&self) {
+        unsafe {
+            pw_thread_loop_signal(self.0.as_ptr(), false);
+        }
+    }
+
+    pub fn wait(&self) {
+        unsafe {
+            pw_thread_loop_wait(self.0.as_ptr());
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
```

Comment:
Clone is wrong here as well, you need to wrap the inner pw_loop with either Arc or Rc.

Reply:
I found it a bit problematic wrapping the inner pw_loop with Arc or Rc, I would leave it with clone, but what is wrong about using clone in this case ? :thinking:

Reply:
> I found it a bit problematic wrapping the inner pw_loop with Arc or Rc,

Which part is problematic ? Because `pw_loop` is not reference counted, see how pipewire-rs handles it https://pipewire.pages.freedesktop.org/pipewire-rs/src/pipewire/main_loop.rs.html#13

Reply:
Here, even after wrapping in Rc they still use clone.

Reply:
You should keep clone yes, but only after you wrapped the inner type with Rc/Arc